internal/state/storage/storage.go

```diff
@@ -0,0 +1,414 @@
+	storage.lastOperationDurationMetric.WithLabelValues(operationSave).Set(duration.Seconds())
+
+	return nil
+}
+
+// Load loads all clusters' states from the persistent storage.
+func (storage *Storage) Load() (err error) {
+	defer func() {
+		if err != nil {
+			storage.operationsMetric.WithLabelValues(operationLoad, statusError).Inc()
+		}
+	}()
+
+	start := time.Now()
+
+	// open file for reading
+	file, err := os.Open(storage.path)
+	if err != nil {
+		return fmt.Errorf("failed to open file: %w", err)
+	}
+
+	defer file.Close() //nolint:errcheck
+
+	stats, err := storage.Import(file)
+	if err != nil {
+		return fmt.Errorf("failed to read snapshot: %w", err)
+	}
+
+	if err = file.Close(); err != nil {
+		return fmt.Errorf("failed to close file: %w", err)
+	}
+
+	duration := time.Since(start)
+
+	storage.logger.Info("state loaded", zap.Int("clusters", stats.NumClusters), zap.Int("affiliates", stats.NumAffiliates),
+		zap.Int("endpoints", stats.NumEndpoints), zap.Duration("duration", duration))
+
+	storage.operationsMetric.WithLabelValues(operationLoad, statusSuccess).Inc()
+	storage.lastSnapshotSizeMetric.WithLabelValues(operationLoad).Set(float64(stats.Size))
+	storage.lastOperationClustersMetric.WithLabelValues(operationLoad).Set(float64(stats.NumClusters))
+	storage.lastOperationAffiliatesMetric.WithLabelValues(operationLoad).Set(float64(stats.NumAffiliates))
+	storage.lastOperationEndpointsMetric.WithLabelValues(operationLoad).Set(float64(stats.NumEndpoints))
+	storage.lastOperationDurationMetric.WithLabelValues(operationLoad).Set(duration.Seconds())
+
+	return nil
+}
+
+// Import imports all clusters' states from the given reader.
+//
+// When importing, we avoid unmarshalling to the storagepb.StateSnapshot type directly, as it causes an allocation of all the cluster snapshots at once.
+// Instead, we process clusters in a streaming manner, unmarshaling them one by one and importing them into the state.
+func (storage *Storage) Import(reader io.Reader) (SnapshotStats, error) {
+	const maxHeaderSize = 20
+
+	size := 0
+	numClusters := 0
+	numAffiliates := 0
+	numEndpoints := 0
+	buffer := make([]byte, 0, maxHeaderSize)
+	reader = bufio.NewReader(reader)
+
+	// unmarshal the clusters in a streaming manner and import them into the state
+	if err := storage.state.ImportClusterSnapshots(func() (*storagepb.ClusterSnapshot, bool, error) {
+		var initialNumRead, additionalNumRead int
+
+		// top-up the buffer to its capacity
+		readStart := len(buffer)
+		buffer = buffer[:cap(buffer)]
+
+		initialNumRead, err := reader.Read(buffer[readStart:])
+		if err != nil && err != io.EOF { //nolint:errorlint
+			return nil, false, fmt.Errorf("failed to read bytes: %w", err)
+		}
+
+		// trim the buffer to the actual read size
+		buffer = buffer[:readStart+initialNumRead]
+
+		if len(buffer) == 0 { // processed all data, return
+			return nil, false, nil
+		}
+
+		headerSize, clusterSize, err := decodeClusterSnapshotHeader(buffer)
```

Comment:
we need to have a maximum supported `clusterSize`

Reply:
Done, also added a test where I create a cluster with the maximum possible marshaled size within the discovery service limits and test importing it when it's just at the max size, and when it is a bit larger than that.

I also use the computed max size from that test as the max allowed value.